### ISSUE TITLE
Fix bug causing missing glyphs

### ISSUE THREAD

**Issue by davidtheclark** [mapbox/mapbox-gl-js#3088]
I'm sorry I'm having some difficulty pinning down exact steps to reproduce, but thought this would still be worth posting. Maybe you can give me some pointers about debugging mapbox-gl-js that would help me find a more repeatable pattern? Anyway, here's the issue ...

I am using Mapbox's satellite+street map as a basemap. I'm noticing that sometimes this map renders at low zooms with some city names clipped. When I zoom in further, the city names are restored to their full glory; but zooming out again shows them clipped, again. I haven't been able to pinpoint a specific behavior that causes this problem. I just know that *once it happens*, it stays that way (within a certain zoom range); though most of the time it doesn't happen.

Here's what I'm seeing — clipped city names:

![screen shot 2016-08-28 at 7 13 56 pm](https://cloud.githubusercontent.com/assets/628431/18039161/05a6220a-6d55-11e6-979e-16b19bb88c31.png)

And, zooming in a little more, the full city names:

![screen shot 2016-08-28 at 7 14 19 pm](https://cloud.githubusercontent.com/assets/628431/18039174/1711a118-6d55-11e6-8124-bcafc0610a68.png)

**Comment by lucaswoj** [comment]
I'm having trouble reproducing the issue too. How often have you seen this behaviour? 

Label clipping has been a symptom of several different faults in the past.

**Comment by davidtheclark** [comment]
I am also having a hard time reproducing it! I see it infrequently, but have seen it frequently enough to notice it: it hasn't happened only once or twice, but several times. Sorry I can't help reproduce better. If you want to close this issue I understand and will try to catch some pattern if I see it again.

**Comment by mb12** [comment]
@davidtheclark  This is a known issue in gl js. It clips symbols and glyphs across tile boundaries.  Set [drawAcrossEdges](https://github.com/mapbox/mapbox-gl-js/blob/c78a89bb69455925d72b029232bbbf7a5bc4652c/js/render/draw_symbol.js#L13) unconditionally to true in your build of gl js as a workaround.

**Comment by jfirebaugh** [comment]
It doesn't look like a clipping issue. Look at the label for "Little Rock", for example. In the bad rendering, it's missing the "L", "R", and "k". It seems highly unlikely that tile clipping could cause letters at the beginning, middle, and end to be clipped.

**Comment by mb12** [comment]
@jfirebaugh You are right. Its unlikely to be the the clipping issue since glyphs in the middle of text are missing.

**Comment by timiyay** [comment]
I'm seeing a similar issue with v0.23.0. I'm on 0.22.1, and exploring the jump to v0.23.0.

Many of the text labels (city, place, road, etc) are majorly corrupted to the point of being unreadable.

A demonstration:

v0.22.1 - renders as expected
![expected](https://cloud.githubusercontent.com/assets/1470250/18132737/46866aa4-6fdb-11e6-8c84-4a58d4ffd47b.png)

v0.23.0 - crazy messed up
![labels_corrupted](https://cloud.githubusercontent.com/assets/1470250/18132747/4df97c90-6fdb-11e6-8b00-5f07729fb00e.png)

The labels seem to be jumbled or corrupted, rather than clipped. This happens across the map, not just at tile boundaries.

**Comment by mourner** [comment]
@timiyay do you have a reproducible JSFiddle test case we could look at?

**Comment by jfirebaugh** [comment]
This and #3086 are likely caused by #2930.

**Comment by lucaswoj** [comment]
I have nowhere to start debugging this issue without reproduction steps. @davidtheclark could you timebox 30 minutes to try to come up with some? 

@timiyay Your issue looks more like  #3086 than this.

**Comment by davidtheclark** [comment]
@lucaswoj I tried playing around for a little while. I can pretty consistently reproduce this on the live demo map, but I was unable to reproduce it in a simplified test case. At https://www.mapbox.com/bites/00288/?nointro=true, try clicking the Beaver Creek Fire just northwest of Denver; about one second into the zoom, shift+double-click to force the zoom to stop; then zoom out. Usually when I do that I see the garbled text. Sometimes I have to repeat it once or twice.

I tried to create a fiddle that similarly zooms so I could try stopping it in in the same way, but could not get the same result. I wonder if it has to do with zooming and transforming the style (showing a previously hidden layer and altering its filter) at the same time.

**Comment by timiyay** [comment]
I will attempt a shareable repro later today (Aussie time), so that we can hone in.

From a quick glance at the related issues and comments, there's 2 possibly-relevant things that our single-page Ember app is doing:

1. calling `map.remove()` when navigating from the map route to a different route, then making a `new mapboxgl.Map()` when returning to the map route
2. toggling the display of various (>40) layers in our Mapbox style, through a combo of toggling `visibility` or `opacity`

@lucaswoj Would you prefer me to move my comments over to #3086, keep them here, or start a new issue?

**Comment by timiyay** [comment]
I've taken an unsuccessful attempt at a repro, which is here: http://codepen.io/timiyay/pen/LRYLjm

I'm using the same Mapbox style, same fonts, etc. I've replicated the basic UX of suspected problem areas:

* navigating away from map and calling `map.remove()`, and then going back again
* toggling transparent layers that overlay the labels

I'll keep trying to repro.

**Comment by lucaswoj** [comment]
@timiyay I think the bug you're seeing is #3086, not the one being discussed in this ticket.

**Comment by jfirebaugh** [comment]
I'm pretty sure this had the same root cause as #3086, which was fixed by #3133 -- closing.
